In [112]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
game_data = pd.read_csv(file_to_load, delimiter=",")
game_data.head(1)

Purchase ID       SN  Age Gender  Item ID  \
0            0  Lisim78   20   Male      108   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53

## Player Count

In [113]:
# total number of players
total_players = len(game_data['SN'].value_counts())
total_players = pd.DataFrame({"Total Players": total_players}, index=[0])
total_players

Total Players
0            576

## Purchasing Analysis (Total)

In [114]:
#output list that is to be appended
fantasy = []
# Number of Unique Items
total_unique = len(game_data['Item ID'].value_counts())
fantasy.append(total_unique)

# Average Purchase Price
avg_purchase_price = game_data["Price"].mean()
fantasy.append('${:,.2f}'.format(avg_purchase_price))

# Total Number of Purchases
total_purchases = len(game_data['Price'])
fantasy.append(total_purchases)

# Total Revenue
total_revenue = game_data["Price"].sum()
fantasy.append('${:,.2f}'.format(total_revenue))
total_revenue

#output
output_purchasing_analysis = pd.DataFrame(columns = ['Number of Unique Items', 'Average Price', 'Number of Purchases',	'Total Revenue'])
output_purchasing_analysis.loc[0] = fantasy
output_purchasing_analysis


Number of Unique Items Average Price Number of Purchases Total Revenue
0                    179         $3.05                 780     $2,379.77

## Gender Demographics

In [115]:
#output list that is to be appended
male = []
female = []
other = []
total_count = len(game_data)

# Percentage and Count of Male Players
gender_group = game_data.groupby("Gender")
male_gender = gender_group.get_group('Male')
male.append(len(male_gender.value_counts()))
male_percentage = '{:,.2f}%'.format(len(male_gender.value_counts())/total_count)
male.append(male_percentage)

# Percentage and Count of Female Players
gender_group = game_data.groupby("Gender")
female_gender = gender_group.get_group('Female')
female.append(len(female_gender.value_counts()))
female_percentage = '{:,.2f}%'.format(len(female_gender.value_counts())/total_count)
female.append(female_percentage)
female_percentage

# Percentage and Count of Other / Non-Disclosed
gender_group = game_data.groupby("Gender")
other_gender = gender_group.get_group('Other / Non-Disclosed')
other.append(len(other_gender.value_counts()))
other_percentage = '{:,.2f}%'.format(len(other_gender.value_counts())/total_count)
other.append(other_percentage)
other_percentage

#output
output_gender_demographics = pd.DataFrame(columns = ['Total Count',	'Percentage of Players'])
output_gender_demographics.rename(index={0:'Male', 1:'Female', 2:'Others / Non-Disclosed'})
output_gender_demographics.loc['Male'] = male
output_gender_demographics.loc['Female'] = female
output_gender_demographics.loc['Others / Non-Disclosed'] = other
output_gender_demographics


Total Count Percentage of Players
Male                           652                 0.84%
Female                         113                 0.14%
Others / Non-Disclosed          15                 0.02%

## Purchasing Analysis (Gender)

In [139]:
#output list that is to be appended
male = []
female = []
other = []

# calculate the average, sum, and average total per person per gender
gender_group = game_data.groupby("Gender")
male_gender = gender_group.get_group('Male')
male.append(len(male_gender))
male.append('${:,.2f}'.format(male_gender['Price'].mean()))
male.append('${:,.2f}'.format(male_gender['Price'].sum()))
man = len(male_gender.value_counts())
male.append('${:,.2f}'.format(male_gender['Price'].sum()/man))

gender_group = game_data.groupby("Gender")
female_gender = gender_group.get_group('Female')
female.append(len(female_gender))
female.append('${:,.2f}'.format(female_gender['Price'].mean()))
female.append('${:,.2f}'.format(female_gender['Price'].sum()))
fem = len(female_gender.value_counts())
female.append('${:,.2f}'.format(female_gender['Price'].sum()/fem))

gender_group = game_data.groupby("Gender")
other_gender = gender_group.get_group('Other / Non-Disclosed')
other.append(len(other_gender))
other.append('${:,.2f}'.format(other_gender['Price'].mean()))
other.append('${:,.2f}'.format(other_gender['Price'].sum()))
o = len(other_gender.value_counts())
other.append('${:,.2f}'.format(other_gender['Price'].sum()/o))

# output
output_gender_purchase_analysis = pd.DataFrame(columns = ['Purchase Count',	'Average Purchase Price',	'Total Purchase Value',	'Avg Total Purchase per Person'])
output_gender_purchase_analysis.rename(index={0:'Male', 1:'Female', 2:'Others / Non-Disclosed'})
output_gender_purchase_analysis.loc['Male'] = male
output_gender_purchase_analysis.loc['Female'] = female
output_gender_purchase_analysis.loc['Others / Non-Disclosed'] = other
output_gender_purchase_analysis

Purchase Count Average Purchase Price  \
Male                              652                  $3.02   
Female                            113                  $3.20   
Others / Non-Disclosed             15                  $3.35   

                       Total Purchase Value Avg Total Purchase per Person  
Male                              $1,967.64                         $3.02  
Female                              $361.94                         $3.20  
Others / Non-Disclosed               $50.19                         $3.35

## Age Demographics

In [117]:
#clear any duplicates
cleaned_game_data = game_data.drop_duplicates('SN')

#create bins and groups
bins = pd.cut(cleaned_game_data['Age'],[0,9,14,19,24,29,34,39,100], labels=["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"])

#function to calculate bins
def age_bins(demo):
    #calculate the total count and the percentage
    column_names = {'Total Count': demo['Age'].count(),
    'Percentage of Players': '{:,.2f}%'.format(demo['Age'].count()*100/cleaned_game_data["Age"].count()),
    }
    return pd.Series(column_names, index=["Total Count", "Percentage of Players"])

#groupby the bins and apply the function
cleaned_game_data.groupby(bins).apply(age_bins)

Total Count Percentage of Players
Age                                     
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

In [141]:
#create bins and groups
bins = pd.cut(game_data['Age'],[0,9,14,19,24,29,34,39,100], labels=["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"])

#function to calculate bins
def avg_price_age(demo):
    #calculate the Purchase count, average purchase price, total purchase value, and average purchase per person
    column_names = {'Purchase Count': demo['Age'].count(),
    'Average Purchase Price': '${:,.2f}'.format(demo['Price'].sum()/demo['Price'].count()),
    'Total Purchase Value': '${:,.2f}'.format(demo['Price'].sum()),
    'Average Purchase per Person': '${:,.2f}'.format(demo['Price'].sum()/demo['SN'].nunique()),
    }
    return pd.Series(column_names, index=['Purchase Count', 'Average Purchase Price', 'Total Purchase Value', 'Average Purchase per Person'])

#groupby the bins and apply the function
game_data.groupby(bins).apply(avg_price_age)

Purchase Count Average Purchase Price Total Purchase Value  \
Age                                                                 
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Average Purchase per Person  
Age                                
<10                         $4.54  
10-14                       $3.76  
15-19                       $3.86  
20-24                       $4.32  
25-29                       $3.81  
30-34                       $4.12  
35-39                       $4.76  
40+                         $3.19

## Top Spenders

In [119]:
#create bins and groups
spenders_group = game_data
spenders_group.groupby('SN')

#function to calculate bins
def top_spenders(demo):

    #calculate the Purchase count, average purchase price, total purchase value, and average purchase per person
    column_names = {'Average Purchase Price': '${:,.2f}'.format(demo['Price'].sum()/demo['Price'].count()),
    'Total Purchase Value': demo['Price'].sum(),
    'Purchase Count': demo['SN'].count(),
    }
    return pd.Series(column_names, index=['Purchase Count', 'Average Purchase Price', 'Total Purchase Value',])

#apply the function to the groupby of spenders_group
spenders_output = spenders_group.groupby(['SN'], sort=False).apply(top_spenders)
spenders_output.nlargest(5, ['Total Purchase Value'])

Purchase Count Average Purchase Price  Total Purchase Value
SN                                                                      
Lisosia93                 5                  $3.79                 18.96
Idastidru52               4                  $3.86                 15.45
Chamjask73                3                  $4.61                 13.83
Iral74                    4                  $3.40                 13.62
Iskadarya95               3                  $4.37                 13.10

## Most Popular Items

In [120]:
items_group = game_data
items_group.groupby(['Item ID', 'Item Name'])

#function to calculate bins
def popular_items(demo):

    #calculate the Purchase count, average purchase price, total purchase value, and average purchase per person
    column_names = {'Purchase Count': demo['Item ID'].count(),
    'Average Item Price': '${:,.2f}'.format(demo['Price'].sum()/demo['Price'].count()),
    'Total Purchase Value': '${:,.2f}'.format(demo['Price'].sum()*demo['Item ID'].count()/demo['Price'].count()),
    }
    return pd.Series(column_names, index=['Purchase Count', 'Average Item Price', 'Total Purchase Value',])

#apply the function to the groupby of items_group
items_output = items_group.groupby(['Item ID', 'Item Name'], sort=False).apply(popular_items)
items_output.nlargest(5, ['Purchase Count'])

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
132     Persuasion                                                 9   

                                                     Average Item Price  \
Item ID Item Name                                                         
92      Final Critic                                              $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm              $4.23   
108     Extraction, Quickblade Of Trembling Hands                 $3.53   
82      Nirvana                                                   $4.90   
132     Persuasion                                                $3.22   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
132     Persuasion                                                 $28.99

## Most Profitable Items

In [173]:
profitable_group = game_data
profitable_group.groupby(['Item ID', 'Item Name'])

#function to calculate bins
def profitable_items(demo):

    #calculate the Purchase count, average purchase price, total purchase value, and average purchase per person
    column_names = {'Purchase Count': demo['Item ID'].count(),
    'Average Item Price': '${:,.2f}'.format(demo['Price'].sum()/demo['Price'].count()),
    'Total Purchase Value': '${:,.2f}'.format(demo['Price'].sum()*demo['Item ID'].count()/demo['Price'].count()),
    }
    return pd.Series(column_names, index=['Purchase Count', 'Average Item Price', 'Total Purchase Value',])

#apply the function to the groupby of profitable_group
items_output = profitable_group.groupby(['Item ID', 'Item Name'], sort=True).apply(profitable_items)
items_output.nlargest(5, ['Purchase Count'])

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
108     Extraction, Quickblade Of Trembling Hands                  9   
132     Persuasion                                                 9   

                                                     Average Item Price  \
Item ID Item Name                                                         
92      Final Critic                                              $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm              $4.23   
82      Nirvana                                                   $4.90   
108     Extraction, Quickblade Of Trembling Hands                 $3.53   
132     Persuasion                                                $3.22   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
132     Persuasion                                                 $28.99

## Data Analysis of Purchase Data for Heroes of Pymoli

Looking at the gender data we see that males consisted of over 84% of the gaming population and had the most purhcases and the highest Total Purchase Value, but Others/Non-Disclosed had the highest Average Total Purchase per Person even though they only consist of 2% of the gaming population. 
The age demographics data showed that the age groups of 20-24 had the highest percentage of players at 44.79%. This age group also had the Highest Total Purchase Value and Purchase Count compared to the other age groups.
For the most popular items and most profitable item data the top 5 most popular items are also the most profitable items. This has a lot to due with the Purchase Count. 